### PR TITLE
Spatial conclusion

### LINKED ISSUE #33: Triple::document redundant
The Triple struct contains 4 values:
- subject
- predicate
- object
and
- document
The latter was intended to contain information of whom this triple belongs to, but since it is stored in an RDFEntity this is redundant, and the SopranoModule e.g. discards it and sets the document-part if the triples it adds to its own model to the id of the object it is extracted from.

Since I ran into the problem that I couldn't remove a Triple because I didn't know the document-part of it, and what the SPARQL Query [1] returned was different from what was stored in the database, I think we should remove "document" from the Triple-struct.

### LINKED ISSUE #55: Boost Geometry and GOES::Geom conflict in Spatialindex
Hi,

currently if you try to add a 2D Geometry to the core with an active SpatialIndex module you will receive following runtime error from the Boost rtree:
`
[ctest] sempr_spatial_conclusion_tests: /usr/include/boost/geometry/index/rtree.hpp:1445: void boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::raw_insert(const value_type&) [with Value = std::pair<boost::geometry::model::box<boost::geometry::model::point<double, 3ul, boost::geometry::cs::cartesian> >, std::shared_ptr<sempr::entity::Geometry> >; Parameters = boost::geometry::index::quadratic<16ul>; IndexableGetter = boost::geometry::index::indexable<std::pair<boost::geometry::model::box<boost::geometry::model::point<double, 3ul, boost::geometry::cs::cartesian> >, std::shared_ptr<sempr::entity::Geometry> > >; EqualTo = boost::geometry::index::equal_to<std::pair<boost::geometry::model::box<boost::geometry::model::point<double, 3ul, boost::geometry::cs::cartesian> >, std::shared_ptr<sempr::entity::Geometry> > >; Allocator = std::allocator<std::pair<boost::geometry::model::box<boost::geometry::model::point<double, 3ul, boost::geometry::cs::cartesian> >, std::shared_ptr<sempr::entity::Geometry> > >; boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::value_type = std::pair<boost::geometry::model::box<boost::geometry::model::point<double, 3ul, boost::geometry::cs::cartesian> >, std::shared_ptr<sempr::entity::Geometry> >]: Assertion `(detail::is_valid(m_members.translator()(value)))&&("Indexable is invalid")' failed.
`

The reason for this is, that geos::geom coordinates stores X, Y, NaN for 2D coordinate and the boost rtree expect valid 3D points without a NaN value.

So how to handle 2D and 3D in the same context? ( set z to 0 instead of NaN)?

### PR DESCRIPTION
**Pull request by ctieben**
Changes:
- Sperate SpatialIndex for 2D and 3D. This will fix #55 
- Add a SpatialConclusion that will generate RDFTriple for geometry relations.
- Also the redudant document store in the triple is changes, fixed #33 
- The RDF Entity could be temporary
- GlobalCS now knows the cardinal direction.
- isEqual now for all SpatialReference and not only for GlobalCS
- Add additional SpatialQuery types
- Add converter for Eigen, Boost and GEOS::GEOM points in the SpatialIndexBase

### CODE REVIEW

File: include/sempr/core/RDF.hpp [hunk 1]
```diff
@@ -3,67 +3,73 @@
...
-    RDFS(Resource)
-    RDFS(Class)
-    RDFS(subClassOf)
-    RDFS(subPropertyOf)
-    RDFS(domain)
-    RDFS(range)
-#   undef RDFS
-}
+        namespace rdfs {
+            const std::string& baseURI();
```
Comment (niniemann) [inline include/sempr/core/RDF.hpp:26]: Why did you add this line? `const std::string& baseURI();` is already declared through the macro below:
```c++
#define RDFS(name) const std::string& (name)();
RDFS(baseURI)
```
Same goes for owl.

File: include/sempr/entity/GeometricObject.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include <sempr/core/RDF.hpp>
+#include <odb/core.hxx>
+
+#include <string>
+
+namespace sempr { namespace entity {
+
+
+#pragma db object
+class GeometricObject : public SemanticEntity {
```
Comment (niniemann) [inline include/sempr/entity/GeometricObject.hpp:16]: Some comments would be nice. What is the purpose of this class?

File: include/sempr/entity/GeometricObject.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    GeometricObject(bool temporary);
+    GeometricObject(const core::IDGenBase*, bool temporary = false);
+
+    const Geometry::Ptr geometry() const { return geometry_; }
+    void geometry(const Geometry::Ptr geometry);
+
+    const std::string type() const { return type_; }
+    void type(const std::string& type);
+
+    void registerProperty(const std::string& predicate, const std::string& object);
```
Comment (niniemann) [inline include/sempr/entity/GeometricObject.hpp:31]: A method of the same name (but templated) already exists in `SemanticEntity`, which you derive from here. Is this on purpose? Or coincidental?

```c++
template <class T> void registerProperty(const std::string& predicate, T& property) { /* ... */ }
// ... and derivations ...
```
These methods are used to register member variables and make them visible in the rdf.

File: include/sempr/entity/RDFVector.hpp [hunk 4]
```diff
@@ -26,20 +22,19 @@ class RDFVectorIterator : public TripleIterator {
...
     RDFVector();
-    RDFVector(const core::IDGenBase*);
+    RDFVector(bool temporary);
+    RDFVector(const core::IDGenBase*, bool temporary = false);
     virtual ~RDFVector(){}
 
     void getTriples(std::vector<Triple>& triples) const;
     const Triple& getTripleAt(const size_t index);
-    bool addTriple(const Triple& triple);
+    bool addTriple(const Triple& triple, bool replace = false);     // if replace than an equal triple will be removed before, otherwise there a multiple entries possible
```
Comment (niniemann) [inline include/sempr/entity/RDFVector.hpp:37]: What's the meaning of "replacing" an equivalent Triple? If there's a way to see a difference they were not equal. So `replace = true` would be the same as `allow_duplicates = false`. The other way round is still intuitive for `allow_duplicates`, but `replace = false` is ambiguous: What happens with an equivalent triple? Is it just discarded without "replacing" the old one, and there is still only one of present in the vector? Or is it added as a duplicate?

tl;dr: I'd suggest renaming `bool replace` to `bool allow_duplicates`.

Comment (niniemann) [reply include/sempr/entity/RDFVector.hpp:37]: Also, maybe the default should be to **not** allow duplicates? I don't see any use case for adding the same fact multiple times.

File: include/sempr/entity/RDFVector.hpp [hunk 5]
```diff
@@ -48,11 +43,15 @@ class RDFVector : public RDFEntity {
     TripleIteratorWrapper begin() const override;
     TripleIteratorWrapper end() const override;
 
+    bool validity(const Triple& triple) const;
```
Comment (niniemann) [inline include/sempr/entity/RDFVector.hpp:46]: What does this check? Does it depend on the RDFVector, or should this be moved to the Triple-class?

File: include/sempr/entity/SemanticEntity.hpp [hunk 6]
```diff
@@ -268,6 +274,13 @@ class SemanticEntity : public RDFEntity {
         properties_.push_back(rprop);
     }
 
+    template <class T>
+    void removeProperty(const std::string& predicate, T& property)
```
Comment (niniemann) [inline include/sempr/entity/SemanticEntity.hpp:278]: Wouldn't `unregisterProperty` be a better name?
May I ask what this is used for? The registerProperty methods were intended to be used on construction of an object, only to export member variables to RDF. Why should you want to unregister such a member?

File: include/sempr/entity/spatial/reference/LocalCS.hpp [hunk 7]
```diff
@@ -41,6 +41,7 @@ class LocalCS : public SpatialReference {
     void setParent(ProjectionCS::Ptr);
     void setParent(GeocentricCS::Ptr);
 
+    bool isEqual(const SpatialReference::Ptr other) const override;
```
Comment (niniemann) [inline include/sempr/entity/spatial/reference/LocalCS.hpp:44]: Just asking: Shouldn't a simple check use `const SpatialReference&`? It's more flexible, since I can always dereference a `shared_ptr`, but creating a `shared_ptr` for a stack variable (e.g.) is not the best idea. ;)

File: include/sempr/processing/SpatialConclusion.hpp [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+                auto selfMax = getMax(selfBox, axis);
+                auto otherMin = getMin(otherBox, axis);
+                return selfMax <= otherMin;
+            }
+        }
+
+        return false;
+    }
+
+    static bool intersectionCheck(const ValuePair& self, const ValuePair& other, const entity::SpatialReference::Ptr& ref)
```
Comment (niniemann) [inline include/sempr/processing/SpatialConclusion.hpp:419]: These methods look like you check all combinations of entities manually, is that correct? Do you use the index / index queries?

File: include/sempr/processing/SpatialIndex.hpp [hunk 9]
```diff
@@ -36,61 +45,426 @@ namespace bgi = boost::geometry::index;
     The bridge between ProjectionCS and GeographicCS could be a bit more difficult as we will need
     coordinate transformations done by GDAL in the process.
 */
-class SpatialIndex
-    : public Module< core::EntityEvent<entity::Geometry>,
+/**
+ * Spatial Index for one specific root coordinate system.
+ * All entities that could not be tranformes to this coordinate system will be skipped!
+ * Note that the SpatialIndex currently only support 3D Indexing. 2D Geometries will be projected to the zero x/y plane.
```
Comment (niniemann) [inline include/sempr/processing/SpatialIndex.hpp:51]: Is this note still correct?

File: include/sempr/processing/SpatialIndex.hpp [hunk 10]
```diff
@@ -36,61 +45,426 @@ namespace bgi = boost::geometry::index;
...
+            if (it->second.second == value.second)
+                return it->first;
+        }
+
+        return nullptr;
+    }
+
+
+    // check if a given pre-set of results also match the geometry based test.
+    void checkGeometry(const ValuePair& ref, SpatialQueryType queryType, std::vector<ValuePair>& results)
```
Comment (niniemann) [inline include/sempr/processing/SpatialIndex.hpp:317]: The checks used here are performed by geos, and thus disregard the Z-dimension, right?
If so, it makes things tricky for some query types (I think especially the negated ones):

If the bounding-box check returns true on a `?a NOT_CONTAINS ?b` check, but the geometry check returns false... What's the truth, now? The projection of the object onto the 2D plane could result in a containment relation that is not present in 3D. So the (3D) BB-check is correct, but the geometry-check still removes the geometry from the result-list. I think it's not possible to have a 3D "NOT_CONTAINS" on the bounding boxes, if the real geometries fulfill a containment relation. So the BB check should be sufficient?

There are definitely a few more edge cases to think about.

tl;dr: 3D bbox check combined with a 2D geometry check is complicated, and not completely correct as it is implemented right now.

File: include/sempr/processing/SpatialIndex.hpp [hunk 11]
```diff
@@ -36,61 +45,426 @@ namespace bgi = boost::geometry::index;
...
+        // this envelope is in the coordinate system of the geometry. But what we need is an envelope
+        // that is axis aligned with the root reference system. We could transform the geometry to root,
+        // take the envelope and transform it back, but that is just ridiculus. Instead: Create a
+        // bounding-box-geometry (8 points, one for each corner), transform it, and take its envelope.
+        //
+        // Note z-coordinate by default is NaN in 2D (defined by geos::geom)
+        // ---
+        // create a geometry with the matching extends: 8 point, every corner must be checked!
+
+        entity::MultiPoint::Ptr mpEntity( new entity::MultiPoint() );    // Note: this wast IDs - recommended to use a factory in future
```
Comment (niniemann) [inline include/sempr/processing/SpatialIndex.hpp:429]: Regarding the note: Yes, correct. We should talk about a strategy for IDs to improve this situation.

File: include/sempr/processing/SpatialIndex.hpp [hunk 12]
```diff
@@ -36,61 +45,426 @@ namespace bgi = boost::geometry::index;
...
+        // transfrom it to the reference system of the R-Tree
+        mpEntity->transformToCS(rootCS_);   //could throw an exception if there is no transformation from geo cs to root cs
+
+        return mpEntity;
+    }
+
+    Box buildBox(entity::Geometry::Ptr geomBox) const
+    {
+        /*  
+        // really strange happenings - this is different to the manually apply of the env filter!!!
```
Comment (niniemann) [inline include/sempr/processing/SpatialIndex.hpp:444]: This should be investigated! A quick look into the source of `Geometry::findEnvelope`:
```c++
void Geometry::findEnvelope(geom::Coordinate& min, geom::Coordinate& max)
{
    EnvelopeFilter envFilter;
    apply(envFilter);   //apply the filter to find the min and max values of this geometry

    min = envFilter.getMin();
    max = envFilter.getMax();
}
// [...]
void Geometry::apply(Filter& filter)
{
    geom::Geometry* geometry = getGeometryMut();   // best match call to the protected getter. If this is non-const you can only call the public getter with a const return value.

    if (!geometry)
        throw TransformException("no geometry to apply a filter!");

    geometry->apply_rw(filter);
}

// [...]
geom::Geometry* Geometry::getGeometryMut()
{
    return nullptr; // have to be override by childs!
}
```

Maybe some child didn't override `getGeometryMut`?

File: src/entity/GeometricObject.cpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    GeometricObject(new core::IDGen<GeometricObject>(), temporary)
+{
+}
+
+GeometricObject::GeometricObject(const core::IDGenBase* idgen, bool temporary) : 
+    SemanticEntity(idgen, temporary)
+{
+    setDiscriminator<GeometricObject>();
+
+    SemanticEntity::registerProperty(sempr::buildURI("subClassOf", sempr::core::rdfs::baseURI()) , "<http://www.opengis.net/ont/geosparql#SpatialObject>");
```
Comment (niniemann) [inline src/entity/GeometricObject.cpp:24]: Why not just use `sempr::core::rdfs::subClassOf()`?

File: src/entity/GeometricObject.cpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+void GeometricObject::geometry(const Geometry::Ptr geometry)
+{
+    if (geometry_)
+        SemanticEntity::removeProperty(sempr::buildURI(geometry_->id(), sempr::baseURI()), sempr::buildURI("subClassOf", sempr::core::rdfs::baseURI()), "<http://www.opengis.net/ont/geosparql#Geometry>");
+
+    geometry_ = geometry;
+
+    if (geometry_)  // cover nullptr set case
+        SemanticEntity::registerProperty(sempr::buildURI(geometry_->id(), sempr::baseURI()), sempr::buildURI("subClassOf", sempr::core::rdfs::baseURI()), "<http://www.opengis.net/ont/geosparql#Geometry>");
```
Comment (niniemann) [inline src/entity/GeometricObject.cpp:35]: This is dangerous!
The `SemanticEntity::registerProperty` expects the object (`T&`) to be a **member variable** of itself. Now you hand it a `const char*` by reference.... :fearful:

What is it that you want to do here? (See previous question: What's the purpose of GeometricObject?)

File: src/entity/GeometricObject.cpp [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    type_ = type;
+
+    SemanticEntity::registerProperty(sempr::buildURI("type", sempr::core::rdf::baseURI()), type_);
+}
+
+void GeometricObject::registerProperty(const std::string& predicate, const std::string& object)
+{
+    std::string obj = object;   // needed! if you try to passthrough object directly you will receive a compiler error based on type missmatch of const&!
+    SemanticEntity::registerProperty(predicate, obj);
```
Comment (niniemann) [inline src/entity/GeometricObject.cpp:50]: :scream: 
When using `registerProperty`, the SemanticEntity keeps the reference to the object given! So the  `obj` must have the same / longer lifetime as the SemanticEntity, but you give it a temporary that runs out of scope right after this method. :fearful:

Comment (niniemann) [reply src/entity/GeometricObject.cpp:50]: (I should definitely try to add a compile-time check that the referenced variable is a member of SemanticEntity. Is this even possible? Through some template magic with decltype etc?)

File: src/entity/GeometricObject.cpp [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+std::set<std::string> GeometricObject::related(const std::string& predicate)
+{
+    std::set<std::string> objects;
+    for(auto t : *this)
+    {   
+        // empty predicate will add all related object to the result list
+        if ( predicate.empty() || t.predicate == predicate)
```
Comment (niniemann) [inline src/entity/GeometricObject.cpp:59]: Again, I don't really understand what this is meant to do. :thinking:
You have triples with an _empty_ predicate stored in this? Why?

File: src/entity/RDFVector.cpp [hunk 17]
```diff
@@ -56,23 +61,36 @@ const Triple& RDFVector::getTripleAt(const size_t index)
...
+        triples_.push_back(triple);
+        //this->changed();
+    }
+
+    return isValid;
+}
+
+bool RDFVector::validity(const Triple& triple) const
 {
+    Triple t = triple;
```
Comment (niniemann) [inline src/entity/RDFVector.cpp:82]: Why the temporary object & copy here?

File: src/entity/SemanticEntity.cpp [hunk 18]
```diff
@@ -29,6 +29,13 @@ std::string RegisteredPropertyBase::predicate() const
     return predicate_;
 }
 
+bool RegisteredPropertyBase::operator==(const RegisteredPropertyBase& other) const
+{
+    return  hasSubject_ == other.hasSubject_ &&
+            subject_ == other.subject_ &&
+            predicate_ == other.predicate_ &&
+            object() == other.object() ;
```
Comment (niniemann) [inline src/entity/SemanticEntity.cpp:37]: This operator only checks if the subjects and predicates equal and the objects **currently** have the same value. To really check if the RegisteredPropertys are equal (i.e., refer to the same member variable) you would need to check the address of the `T& value_` in the derived class. But that's obviously not so easy. And I still think this is born from the desire to misuse the SemanticEntity. :wink:

File: src/processing/CalculatePointsModule.cpp [hunk 19]
```diff
@@ -32,23 +32,23 @@ void CalculatePointsModule::process(query::CalculatePointsQuery::Ptr query)
...
-    std::static_pointer_cast<entity::MultiPoint>(spatial->refGeo())->setCoordinates(coords);
+    std::static_pointer_cast<entity::MultiPoint>(spatial->refBoxGeometryPair().second)->setCoordinates(coords);
+
     ask(spatial);
     entity::PointCloud::Ptr entity = entity::PointCloud::Ptr(new entity::PointCloud());
 
     for(auto& c : spatial->results)
     {
-        if(c->discriminator() == "sempr::entity::PointCloud")
+        if(c->discriminator() == "sempr::entity::PointCloud")   // could this be done by the dynmaic cast?
```
Comment (niniemann) [inline src/processing/CalculatePointsModule.cpp:63]: A nicer version would be to not hardcode the discriminator, but use the odb type traits:

```c++
if (c->discriminator() == odb::object_traits_impl<sempr::entity::PointCloud, odb::id_common>::info.discriminator)
```
should work. (Though this is still a check for **exactly** PointClouds. This does **not** match any derived entities. Just sayin').

File: test/main.cpp [hunk 20]
```diff
@@ -85,7 +85,7 @@ int main(int argc, char** args)
     DebugModule::Ptr debug(new DebugModule());
     ActiveObjectStore::Ptr active( new ActiveObjectStore() );
     SopranoModule::Ptr semantic( new SopranoModule() );
-    SpatialIndex::Ptr spatial( new SpatialIndex() );
+    SpatialIndex3D::Ptr spatial( new SpatialIndex3D(std::make_shared<LocalCS>()) );
```
Comment (niniemann) [inline test/main.cpp:88]: How can this work? I thought entities need to be located relative to the coordinate system given to the index, in order for the index to work?

**Review by niniemann — CHANGES_REQUESTED** [review]
Phew, a lot has happened since I last looked into sempr. :sweat_smile:
There are some things that definitely should be addressed:

- [ ] The whole point of GeometricObject is not yet clear to me
- [ ] registerProperty has been abused by the GeometricObject beyond all good and evil
- [ ] The strange behaviour of findEnvelope you mentioned should be investigated
- [ ] The geometry checks in the SpatialIndex might expose unintuitive/unwanted behaviour in the 3D case (because the "more precise" geometry checks are 2D)

Plus some minor, less important remarks, but those should be looked into as well. :)
Nevertheless, thanks for all the contributions, and sorry for the late review!

**Comment by niniemann** [comment]
More than stale. (Development proceeded in an internal git).
